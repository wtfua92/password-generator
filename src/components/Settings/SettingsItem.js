import React from "react";
import PropTypes from "prop-types";

function SettingsItem({ value, onChangeHandler, type }) {
  return (
    <div className="password-generator__settings__item">
      <label htmlFor={`setting-item-${type}`}>
        <input
          id={`setting-item-${type}`}
          type="checkbox"
          value={value}
          onChange={() => onChangeHandler(type)}
        />
        <span>use {type}</span>
      </label>
    </div>
  );
}

SettingsItem.defaultProps = {
  value: false,
  onChangeHandler: () => {},
  type: ""
};

SettingsItem.propTypes = {
  value: PropTypes.bool,
  onChangeHandler: PropTypes.func,
  type: PropTypes.string
};

export default SettingsItem;
