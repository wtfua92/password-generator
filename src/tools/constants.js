// min max
// A - Z
export const charRange = [65, 90];
// 0 - 9
export const numRange = [0, 9];
// ! # $ % & ( ) * + - / = @ \ ^ _
export const symbols = [33, 35, 36, 37, 38, 40, 41, 42, 43, 45, 47, 61, 64, 92, 94, 95];

export const regExps = {
    CHAR: /[A-z]/,
    SYMBOL: /[!#$%&()*+\-/=@\\^_]/,
    NUMBER: /\d/,
    UPPERCASE: /[A-Z]/
};

export const types = {
    NUMBER: 'NUMBER',
    SYMBOL: 'SYMBOL',
    UPPERCASE: 'UPPERCASE',
    CHAR: 'CHAR'
};

export default {
    charRange,
    numRange,
    symbols,
    types,
    regExps
};