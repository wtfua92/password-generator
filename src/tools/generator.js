import { types } from "./constants";

class PasswordGenerator {
  constructor() {
    this.alphabeticValues = "abcdefghijklmnopqrstuvwxyz";
    this[types.UPPERCASE] = this.alphabeticValues.toUpperCase();
    this[types.INTEGER] = "0123456789";
    this[types.SYMBOL] = "!@#$%^&*()";
  }

  getRandomNumber(min, max) {
    min = Math.ceil(min);
    max = Math.floor(max);
    return Math.floor(Math.random() * (max - min + 1)) + min;
  }

  getRandomValue(values) {
    return values[this.getRandomNumber(0, values.length - 1)];
  }

  passwordGeneratorValues(valueType) {
    return this.alphabeticValues + this[valueType];
  }

  passwordMeetsRequirements(password, settings = {}) {
    let validPassword = [];
    for (let setting in settings) {
      if (settings[setting]) {
        const regex = new RegExp(`[${this[setting]}]`);
        validPassword.push(password.search(regex) >= 0);
      }
    }
    return !validPassword.includes(false);
  }

  generatePassword(passwordLength = 8, passwordSettings = {}) {
    let password = "";
    let values = this.alphabeticValues;

    for (let setting in passwordSettings) {
      if (passwordSettings[setting]) {
        values += this.passwordGeneratorValues(setting);
      }
    }

    for (let i = 0; i < passwordLength; i++) {
      password += this.getRandomValue(values);
    }

    if (!this.passwordMeetsRequirements(password, passwordSettings)) {
      password = this.generatePassword(passwordLength, passwordSettings);
    }

    return password;
  }
}

export default PasswordGenerator;
