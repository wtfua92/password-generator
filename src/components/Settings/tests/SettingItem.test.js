import React from "react";
import { render, fireEvent, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import SettingsItem from "../SettingsItem";
import { types } from "../../../tools/constants";

describe("<PasswordLength />", function() {
  const onChangeHandlerMock = jest.fn();
  const props = {
    value: false,
    type: types.INTEGER,
    onChangeHandler: onChangeHandlerMock
  };

  let container;

  beforeEach(() => {
    container = render(<SettingsItem {...props} />);
  });

  afterEach(cleanup);

  it("should render successfully if all necessary props are provided", () => {
    const { getByText, getByLabelText } = container;
    const spanElement = getByText(`use ${props.type}`);
    const inputElement = getByLabelText(`use ${props.type}`);
    expect(spanElement).toBeDefined();
    expect(inputElement).toBeDefined();
    expect(inputElement.value).toBe(props.value.toString());
  });

  it("should call onChangeHandler on click", () => {
    const { getByLabelText } = container;
    const inputElement = getByLabelText(`use ${props.type}`);
    fireEvent.click(inputElement, { target: { value: true } });
    expect(onChangeHandlerMock).toHaveBeenCalledWith(props.type);
  });

  it("should match snapshot", () => {
    const {
      container: { firstChild: root }
    } = container;
    expect(root).toMatchSnapshot();
  });
});
