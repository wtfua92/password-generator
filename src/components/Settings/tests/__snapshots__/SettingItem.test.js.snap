// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SettingsItem /> should match snapshot 1`] = `
<div
  class="password-generator__settings__item"
>
  <label
    for="setting-item-integer"
  >
    <input
      id="setting-item-integer"
      type="checkbox"
      value="false"
    />
    <span>
      use 
      integer
    </span>
  </label>
</div>
`;
